File: deno/src/tunnel.ts
import { join } from "jsr:@std/path@1";

const CACHE_DIR = join(Deno.env.get("HOME") ?? Deno.env.get("USERPROFILE") ?? ".", ".porthole", "bin");

function getDownloadUrl(): { url: string; archive: boolean } {
  const os = Deno.build.os;
  const arch = Deno.build.arch === "aarch64" ? "arm64" : "amd64";

  if (os === "darwin") {
    return {
      url: `https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-darwin-${arch}.tgz`,
      archive: true,
    };
  }
  if (os === "linux") {
    return {
      url: `https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-${arch}`,
      archive: false,
    };
  }
  if (os === "windows") {
    return {
      url: `https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-windows-${arch}.exe`,
      archive: false,
    };
  }
  throw new Error(`Unsupported platform: ${os}/${arch}`);
}

function binaryName(): string {
  return Deno.build.os === "windows" ? "cloudflared.exe" : "cloudflared";
}

/** Check if cloudflared is available in PATH */
async function findInPath(): Promise<string | null> {
  try {
    const cmd = new Deno.Command("cloudflared", {
      args: ["--version"],
      stdout: "null",
      stderr: "null",
    });
    const { success } = await cmd.output();
    return success ? "cloudflared" : null;
  } catch {
    return null;
  }
}

/** Download and cache the cloudflared binary, returns the path */
async function downloadCloudflared(): Promise<string> {
  const binPath = join(CACHE_DIR, binaryName());

  // Already cached
  try {
    await Deno.stat(binPath);
    return binPath;
  } catch {
    // not cached yet
  }

  const { url, archive } = getDownloadUrl();

  console.log(`Downloading cloudflared from ${url}...`);
  await Deno.mkdir(CACHE_DIR, { recursive: true });

  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to download cloudflared: ${res.status} ${res.statusText}`);
  }

  if (archive) {
    // macOS .tgz — extract using tar
    const tmpTgz = join(CACHE_DIR, "cloudflared.tgz");
    const body = await res.arrayBuffer();
    await Deno.writeFile(tmpTgz, new Uint8Array(body));

    const tar = new Deno.Command("tar", {
      args: ["-xzf", tmpTgz, "-C", CACHE_DIR],
      stdout: "null",
      stderr: "piped",
    });
    const { success, stderr } = await tar.output();
    await Deno.remove(tmpTgz).catch(() => {});
    if (!success) {
      const msg = new TextDecoder().decode(stderr);
      throw new Error(`Failed to extract cloudflared: ${msg}`);
    }
  } else {
    // Linux / Windows — direct binary
    const body = await res.arrayBuffer();
    await Deno.writeFile(binPath, new Uint8Array(body));
  }

  // Make executable on unix
  if (Deno.build.os !== "windows") {
    await Deno.chmod(binPath, 0o755);
  }

  console.log(`cloudflared cached at ${binPath}`);
  return binPath;
}

/** Ensure cloudflared is available — returns the path to the binary */
export async function ensureCloudflared(): Promise<string> {
  // Prefer system-installed version
  const systemPath = await findInPath();
  if (systemPath) return systemPath;

  // Auto-download to cache
  return await downloadCloudflared();
}

export interface TunnelHandle {
  url: string;
  close: () => void;
}

export async function openTunnel(port: number): Promise<TunnelHandle> {
  const cloudflaredBin = await ensureCloudflared();

  const cmd = new Deno.Command(cloudflaredBin, {
    args: ["tunnel", "--url", `http://localhost:${port}`, "--no-autoupdate"],
    stdout: "piped",
    stderr: "piped",
  });

  const process = cmd.spawn();
  const url = await extractTunnelUrl(process.stderr);

  return {
    url,
    close() {
      try {
        process.kill("SIGTERM");
      } catch {
        // already dead
      }
    },
  };
}

async function extractTunnelUrl(stderr: ReadableStream<Uint8Array>): Promise<string> {
  const reader = stderr.getReader();
  const decoder = new TextDecoder();
  const urlPattern = /https:\/\/[a-z0-9-]+\.trycloudflare\.com/;

  const timeout = AbortSignal.timeout(30_000);
  let buffer = "";

  try {
    while (!timeout.aborted) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const match = buffer.match(urlPattern);
      if (match) {
        reader.releaseLock();
        return match[0];
      }
    }
  } catch {
    // timeout or read error
  }

  reader.releaseLock();
  throw new Error("Failed to extract tunnel URL from cloudflared output within 30s");
}

File: python/src/porthole/tunnel.py
"""Cloudflare Quick Tunnel integration."""

from __future__ import annotations

import asyncio
import os
import platform
import re
import shutil
import stat
import subprocess
import tarfile
from pathlib import Path


CACHE_DIR = Path.home() / ".porthole" / "bin"
URL_PATTERN = re.compile(r"https://[a-z0-9-]+\.trycloudflare\.com")


def _get_download_url() -> tuple[str, bool]:
    """Returns (url, is_archive)."""
    os_name = platform.system().lower()
    cpu = "arm64" if platform.machine() in ("arm64", "aarch64") else "amd64"

    if os_name == "darwin":
        return (
            f"https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-darwin-{cpu}.tgz",
            True,
        )
    if os_name == "linux":
        return (
            f"https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-{cpu}",
            False,
        )
    if os_name == "windows":
        return (
            f"https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-windows-{cpu}.exe",
            False,
        )
    raise RuntimeError(f"Unsupported platform: {os_name}/{cpu}")


def _binary_name() -> str:
    return "cloudflared.exe" if platform.system().lower() == "windows" else "cloudflared"


def _find_in_path() -> str | None:
    return shutil.which("cloudflared")


async def _download_cloudflared() -> str:
    bin_path = CACHE_DIR / _binary_name()

    if bin_path.exists():
        return str(bin_path)

    url, is_archive = _get_download_url()
    print(f"Downloading cloudflared from {url}...")
    CACHE_DIR.mkdir(parents=True, exist_ok=True)

    import aiohttp

    async with aiohttp.ClientSession() as session:
        async with session.get(url) as resp:
            if resp.status != 200:
                raise RuntimeError(f"Failed to download cloudflared: {resp.status}")
            data = await resp.read()

    if is_archive:
        tmp_tgz = CACHE_DIR / "cloudflared.tgz"
        tmp_tgz.write_bytes(data)
        with tarfile.open(tmp_tgz, "r:gz") as tar:
            tar.extractall(path=CACHE_DIR)
        tmp_tgz.unlink(missing_ok=True)
    else:
        bin_path.write_bytes(data)

    if platform.system().lower() != "windows":
        bin_path.chmod(bin_path.stat().st_mode | stat.S_IEXEC)

    print(f"cloudflared cached at {bin_path}")
    return str(bin_path)


async def ensure_cloudflared() -> str:
    system_path = _find_in_path()
    if system_path:
        return system_path
    return await _download_cloudflared()


class TunnelHandle:
    def __init__(self, url: str, process: asyncio.subprocess.Process):
        self.url = url
        self._process = process

    def close(self) -> None:
        try:
            self._process.terminate()
        except ProcessLookupError:
            pass


async def open_tunnel(port: int) -> TunnelHandle:
    cloudflared_bin = await ensure_cloudflared()

    process = await asyncio.create_subprocess_exec(
        cloudflared_bin,
        "tunnel",
        "--url",
        f"http://localhost:{port}",
        "--no-autoupdate",
        stdin=asyncio.subprocess.DEVNULL,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    url = await _extract_tunnel_url(process.stderr)
    return TunnelHandle(url=url, process=process)


async def _extract_tunnel_url(stderr: asyncio.StreamReader) -> str:
    buffer = ""
    try:
        async with asyncio.timeout(30):
            while True:
                chunk = await stderr.read(4096)
                if not chunk:
                    break
                buffer += chunk.decode("utf-8", errors="replace")
                match = URL_PATTERN.search(buffer)
                if match:
                    return match.group(0)
    except asyncio.TimeoutError:
        pass
    raise RuntimeError("Failed to extract tunnel URL from cloudflared output within 30s")
